File: src/h1tdfsdiameter/Java/com/example/tdfsdiameter/BinaryTree.java
package com.example.tdfsdiameter;

import java.util.*;

public class BinaryTree<T> {
    public TreeNode<T> root;

    public BinaryTree(List<TreeNode<T>> listOfNodes) {
        root = createBinaryTree(listOfNodes);
    }

    private TreeNode<T> createBinaryTree(List<TreeNode<T>> listOfNodes) {
        if (listOfNodes == null || listOfNodes.isEmpty()) {
            return null;
        }

        TreeNode<T> root = listOfNodes.get(0);
        Queue<TreeNode<T>> queue = new LinkedList<>();
        queue.offer(root);

        int i = 1;
        while (i < listOfNodes.size()) {
            TreeNode<T> current = queue.poll();

            if (i < listOfNodes.size() && listOfNodes.get(i) != null) {
                current.left = listOfNodes.get(i);
                queue.offer(current.left);
            }
            i++;

            if (i < listOfNodes.size() && listOfNodes.get(i) != null) {
                current.right = listOfNodes.get(i);
                queue.offer(current.right);
            }
            i++;
        }

        return root;
    }

    private int diameterHelper(TreeNode<T> node, int[] diameter) {
        if (node == null) {
            return 0;
        }

        int leftHeight = diameterHelper(node.left, diameter);
        int rightHeight = diameterHelper(node.right, diameter);

        diameter[0] = Math.max(diameter[0], leftHeight + rightHeight);

        return Math.max(leftHeight, rightHeight) + 1;
    }

    public int diameterOfBinaryTree() {
        int[] diameter = new int[1];
        diameterHelper(root, diameter);
        return diameter[0];
    }
}

File: src/h1tdfsdiameter/CS/TDFSDiameter.Algo/BinaryTree.cs
// File: BinaryTree.cs

namespace TDFSDiameter.Algo;

public class BinaryTree<T>(List<TreeNode<T>?> listOfNodes)
{
    public TreeNode<T>? Root { get; private set; } = BinaryTree<T>.CreateBinaryTree(listOfNodes);

    private static TreeNode<T>? CreateBinaryTree(List<TreeNode<T>?> listOfNodes)
    {
        if (listOfNodes == null || listOfNodes.Count == 0 || listOfNodes[0] == null)
        {
            return null;
        }

        TreeNode<T>? root = listOfNodes[0];
        Queue<TreeNode<T>> queue = new();
        queue.Enqueue(root!);

        int i = 1;
        while (i < listOfNodes.Count)
        {
            TreeNode<T>? current = queue.Dequeue();

            if (i < listOfNodes.Count && listOfNodes[i] != null)
            {
                current.Left = listOfNodes[i];
                queue.Enqueue(current.Left!);
            }
            i++;

            if (i < listOfNodes.Count && listOfNodes[i] != null)
            {
                current.Right = listOfNodes[i];
                queue.Enqueue(current.Right!);
            }
            i++;
        }

        return root;
    }

    private static int DiameterHelper(TreeNode<T>? node, ref int diameter)
    {
        if (node == null)
        {
            return 0;
        }

        int leftHeight = BinaryTree<T>.DiameterHelper(node.Left, ref diameter);
        int rightHeight = BinaryTree<T>.DiameterHelper(node.Right, ref diameter);

        diameter = Math.Max(diameter, leftHeight + rightHeight);

        return Math.Max(leftHeight, rightHeight) + 1;
    }

    public int DiameterOfBinaryTree()
    {
        int diameter = 0;

        if (Root != null)
        {
            BinaryTree<T>.DiameterHelper(Root, ref diameter);
        }

        return diameter;
    }
}
